src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -1339,6 +1357,54 @@ public ActivityIdFormat IdFormat
             private set => _state = (_state & ~State.FormatFlags) | (State)((byte)value & (byte)State.FormatFlags);
         }
 
+        /// <summary>
+        /// Enumerates the data stored on an Activity object.
+        /// </summary>
+        /// <typeparam name="T">Type being enumerated.</typeparam>
+        public struct Enumerator<T>
+        {
+            private static readonly DiagNode<T> s_Empty = new DiagNode<T>(default!);
+
+            private DiagNode<T>? _nextNode;
+            private DiagNode<T> _currentNode;
+
+            internal Enumerator(DiagNode<T>? head)
```

Comment:
>DiagNode<T>? head [](http://example.com/codeflow?start=32&length=17)

Is it going to cause allocation if we pass IEumerable<T> instead of passing DiagNode<T>? The enumerators we will use at that time can have ref T Current still.

Reply:
I feel like it might be best to leave as-is. But here are some thoughts, LMK what you think...

* `IEnumerable<T>`:

I don't think this is workable. It would cause allocations and I don't think we can use the `readonly ref`. Basically this would undo everything we are trying to gain 😄 

```csharp
        public struct Enumerator<T>
        {
            private readonly IEnumerator<T> _enumerator;

            internal Enumerator(IEnumerable<T> source)
            {
                _enumerator = source.GetEnumerator(); // If underlying enumerator is a struct, this will box it to IEnumerator<T> (allocation).
            }

            public readonly Enumerator<T> GetEnumerator() => this;

            public readonly ref T Current => ref _enumerator.Current; // Doesn't compile. IEnumerator<T>.Current would also need to be readonly ref?

            public bool MoveNext()
            {
                return _enumerator.MoveNext();
            }
        }
```

* `DiagLinkedList`:

Only issue with this is tags don't use `DiagLinkedList` they have their own [TagsLinkedList](https://github.com/dotnet/runtime/blob/cfd1241efc7ea54a00db848c6d00cd773ed6ae50/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1424).

Code would have to be something like this but compiler is complaining because `TagsLinkedList` isn't a generic thing.

```csharp
        public struct Enumerator<T>
        {
            private DiagEnumerator<T> _enumerator;

            internal Enumerator(DiagLinkedList<T> source)
            {
                _enumerator = source.GetEnumerator();
            }

            internal Enumerator(TagsLinkedList source)
            {
                _enumerator = source.GetEnumerator(); // Doesn't compile. TagsLinkedList isn't generic so the <T> is breaking.
            }

            public readonly Enumerator<T> GetEnumerator() => this;

            public readonly ref T Current => ref _enumerator.Current; // DiagEnumerator also needs readonly ref for this to work. but that isn't a big deal.

            public bool MoveNext()
            {
                return _enumerator.MoveNext();
            }
        }
```

If we could somehow solve the generic thing, this might work. Could be more code to solve that than just passing in `_head` though. Perf could also be impacted slightly due to nested call/virtcall but not 100% sure about that.

* `DiagEnumerator<T>`:

We could pass in a `DiagEnumerator<T>` and just wrap it. We are taking a bit of a perf hit here, the `DiagEnumerator<T>` is going to be copied a couple times or once if we pass by in/ref. But what we lose here is the nice "null-means-empty" behavior of the current implementation. What we would have to do in the `Enumerate*` methods is detect null/empty and hand back an empty enumerator if needed. That will be slower than the current implementation which doesn't need to differentiate between empty and non-empty sources it just passes in valid head or null.

```csharp
        public struct Enumerator<T>
        {
            private DiagEnumerator<T> _enumerator;

            internal Enumerator(DiagEnumerator<T> enumerator)
            {
                _enumerator = enumerator;
            }

            public readonly Enumerator<T> GetEnumerator() => this;

            public readonly ref T Current => ref _enumerator.Current; // Need to update DiagEnumerator<T> to use readonly ref for this to work, but not a big deal to do that I don't think.

            public bool MoveNext()
            {
                return _enumerator.MoveNext();
            }
        }
```

Not opposed to this. But could end up being more code to handle the empty/null cases. And slightly slower.

Reply:
I agree with you then. Let's keep what you have as it will be simpler.